plugins/woocommerce/src/Internal/DataStores/Orders/FeatureController.php

```diff
@@ -0,0 +1,324 @@
+			999
+		);
+
+		add_filter(
+			'woocommerce_debug_tools',
+			function( $tools ) {
+				return $this->add_initiate_regeneration_entry_to_tools_array( $tools );
+			},
+			1,
+			999
+		);
+
+		add_filter(
+			'woocommerce_get_sections_advanced',
+			function( $sections ) {
+				return $this->get_settings_sections( $sections );
+			},
+			1,
+			999
+		);
+
+		add_filter(
+			'woocommerce_get_settings_advanced',
+			function ( $settings, $section_id ) {
+				return $this->get_settings( $settings, $section_id );
+			},
+			2,
+			999
+		);
+	}
+
+	/**
+	 * Class initialization, invoked by the DI container.
+	 *
+	 * @internal
+	 * @param OrdersTableDataStore $data_store The data store to use.
+	 * @param DataSynchronizer     $data_synchronizer The data synchronizer to use.
+	 */
+	final public function init( OrdersTableDataStore $data_store, DataSynchronizer $data_synchronizer ) {
+		$this->data_store        = $data_store;
+		$this->data_synchronizer = $data_synchronizer;
+	}
+
+	/**
+	 * Checks if the feature is visible (so that dedicated entries will be added to the debug tools page).
+	 *
+	 * @return bool True if the feature is visible.
+	 */
+	public function is_feature_visible(): bool {
+		return $this->is_feature_visible;
+	}
+
+	/**
+	 * Makes the feature visible, so that dedicated entries will be added to the debug tools page.
+	 */
+	public function show_feature() {
+		$this->is_feature_visible = true;
+	}
+
+	/**
+	 * Hides the feature, so that no entries will be added to the debug tools page.
+	 */
+	public function hide_feature() {
+		$this->is_feature_visible = false;
+	}
+
+	/**
+	 * Is the custom orders table usage enabled via settings?
+	 * This can be true only if the feature is enabled and a table regeneration has been completed.
+	 *
+	 * @return bool True if the custom orders table usage is enabled
+	 */
+	public function custom_orders_table_usage_is_enabled(): bool {
+		return 'yes' === get_option( self::CUSTOM_ORDERS_TABLE_USAGE_ENABLED_OPTION );
+	}
+
+	/**
+	 * Gets the instance of the orders data store to use.
+	 *
+	 * @param WC_Object_Data_Store_Interface|string $default_data_store The default data store (as received via the woocommerce_order_data_store hooks).
+	 * @return WC_Object_Data_Store_Interface|string The actual data store to use.
+	 */
+	private function get_data_store_instance( $default_data_store ) {
+		if ( $this->is_feature_visible() && $this->custom_orders_table_usage_is_enabled() && ! $this->data_synchronizer->data_regeneration_is_in_progress() ) {
+			return $this->data_store;
+		} else {
+			return $default_data_store;
+		}
+	}
+
+	/**
+	 * Add an entry to Status - Tools to create or regenerate the custom orders table,
+	 * and also an entry to delete the table as appropriate.
+	 *
+	 * @param array $tools_array The array of tools to add the tool to.
+	 * @return array The updated array of tools-
+	 */
+	private function add_initiate_regeneration_entry_to_tools_array(array $tools_array ): array {
```

Comment:
shouldn't this be provided by data synchroniser class since it's tied to how syncing data?

Reply:
This is like that on purpose. I have "designed" the classes so that `FeatureController` takes care of all the plumbing and this includes providing the UI components, the data sync class on the other hand is "headless" and only deals with whatever data handling that isn't covered by the data store class.